### DISCUSSION TITLE
Does @react-native-firebase/app currently support FCM(i.e.@react-native-firebase/messaging) now?

### DISCUSSION THREAD

**Discussion by arlovip** [invertase/react-native-firebase#4266] [Q&A] [answered]
Thanks for this great lib first. I'm using this lib to implement `FCM` targeted on Android. But I come to some problems with this. The dependencies I'm using as follows:

```
  "dependencies": {
    "@react-native-firebase/app": "^8.4.2",
    "@react-native-firebase/messaging": "^7.8.6",
    "react": "16.13.1",
    "react-native": "0.63.2",
    ...others
  }
```

And I see the instructions about [Core/App](https://rnfirebase.io/app/usage#secondary-apps) following the official docs, the screenshot:

![screenshot](https://user-images.githubusercontent.com/22851821/93417995-ea42e400-f8db-11ea-9b17-bb687696d09e.png)

The screenshot above seems to show that `@react-native-firebase/app` doest not support `@react-native-firebase/messaging`. Or I get it wrong?

Sorry, I'm unfamiliar with `Android` platform. What I'd like to ask is:  Can I implement `FCM` with `@react-native-firebase/app` and `@react-native-firebase/messaging` now? How do I do that? I can't find any info about latest versions. 

Any help is appreciated so much.

**Comment by mikehardy** [comment, marked as answer]
FCM works fine, you are looking at which modules support *secondary* projects. Like you could have 2 or more firebase projects, and in the same react-native app initialize 2 or more remote-config modules or similar

That's not what you want

Just read the guide for installing /app and /messaging and it's fine

**Reply by arlovip** [reply]
@mikehardy Thank you for quick reply. Maybe I miss something with that. I'll check it and try to integrate them into my project following the docs guide again.

**Reply by phanithken** [reply]
@lchenfox Would you mind telling me how did you use FCM in secondary project ?
I am running into to in issue where secondary app cannot be used in`@react-native-firebase/messaging` ?

```
Error: You attempted to call "firebase.app('MY_SECONDARY_APP').messaging" but; messaging does not support multiple Firebase Apps.
Ensure you access messaging from the default application only.
```

**Reply by mikehardy** [reply]
@phanithken I do not think @lchenfox was trying to do messaging with more than one app in their app, per my comment. What you want is different, and not supported, likely  not possible.

It's not that messaging is not multi-app in react-native-firebase, it's that messaging is not multi-app in firebase-android-sdk - it is just *not* supported at the native level: https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/FirebaseMessaging#public-static-synchronized-firebasemessaging-getinstance (note, no getInstance that takes a FirebaseApp as a parameter)

Why? Because messaging wakes your app up via some system services - since it is doing app init from messaging and not your normal pathway without the ability for most of your app to run before a notification is potentially displayed, your app has no chance to programmatically install a non-default configuration.

**Reply by phanithken** [reply]
@mikehardy Thank you for quick response. I can understand the reason why it's impossible. In my case, I am developing an application where a firebase app config can be obtained from a distributed API. Using react-native-firebase, It's required me to embed a google-services.json to create the app bundle. so I created an initial firebase project just for the sake of app bundling. After user login, a dedicated firebase configuration for that login user will be return as response from the API and it's initialized in react-native application as secondary app. I want to use firebase messaging on the account returned from the login API. I was thinking of two ways. First one is to build the application without google-services.json and then initialize the **DEFAULT** firebase project at runtime. Second is to change the DEFAULT firebase app after user login complete. I had these ideas but for the time being I don't know how to proceed yet. If you know the procedure or good ideas, would you mind sharing it with me ? 
Apologize me for my poor knowledge!

![IMG_0008](https://user-images.githubusercontent.com/33407061/149373440-98ec8633-8933-4099-b927-b41d799807fa.jpg)

**Reply by mikehardy** [reply]
I can't think of a way that would work without implementing / overriding the native implementions provided by underlying SDKs. I would pursue this upstream in android (and iOS....) SDK repositories, searching there then asking if it is not already addressed

**Reply by phanithken** [reply]
@mikehardy I understood. Let's me take it into a little future consideration. Thank you for your help!
